servicemanager/servicemanager_test.go

```diff
@@ -0,0 +1,605 @@
+
+type testServiceStorage struct {
+	Services []servicemanager.ServiceInfo
+}
+
+/**********************************************************************************************************************
+ * Vars
+ *********************************************************************************************************************/
+
+var tmpDir string
+
+/**********************************************************************************************************************
+ * Init
+ *********************************************************************************************************************/
+
+func init() {
+	log.SetFormatter(&log.TextFormatter{
+		DisableTimestamp: false,
+		TimestampFormat:  "2006-01-02 15:04:05.000",
+		FullTimestamp:    true,
+	})
+	log.SetLevel(log.DebugLevel)
+	log.SetOutput(os.Stdout)
+}
+
+/**********************************************************************************************************************
+* Main
+**********************************************************************************************************************/
+
+func TestMain(m *testing.M) {
+	if err := setup(); err != nil {
+		log.Fatalf("Error setting up: %s", err)
+	}
+
+	ret := m.Run()
+
+	cleanup()
+
+	os.Exit(ret)
+}
+
+/**********************************************************************************************************************
+* Tests
+**********************************************************************************************************************/
+
+func TestInstallService(t *testing.T) {
+	serviceStorage := &testServiceStorage{
+		Services: []servicemanager.ServiceInfo{
+			{ServiceID: "id1", AosVersion: 1, GID: 5000},
+			{ServiceID: "id2", AosVersion: 1, GID: 5001},
+		},
+	}
+
+	config := &config.Config{
+		WorkingDir:  tmpDir,
+		ServicesDir: path.Join(tmpDir, "servicemanager", "services"),
+		DownloadDir: path.Join(tmpDir, "downloads"),
+	}
+
+	if _, err := servicemanager.New(config, serviceStorage); err != nil {
+		t.Fatalf("Can't create SM: %s", err)
+	}
+
+	sm, err := servicemanager.New(config, serviceStorage)
+	if err != nil {
+		t.Fatalf("Can't create SM: %s", err)
+	}
+
+	// install services
+	serviceID := "testService0"
+
+	serviceURL, fileInfo, err := prepareService("Service content")
+	if err != nil {
+		t.Fatalf("Can't prepare test service: %s", err)
+	}
+
+	if err = sm.InstallService(servicemanager.ServiceInfo{ServiceID: serviceID, AosVersion: 1},
+		serviceURL, fileInfo); err != nil {
+		t.Errorf("Can't install service: %s", err)
+	}
+
+	// update service
+	if err = sm.InstallService(servicemanager.ServiceInfo{ServiceID: serviceID, AosVersion: 2},
+		serviceURL, fileInfo); err != nil {
+		t.Errorf("Can't update service: %s", err)
+	}
+
+	os.RemoveAll(serviceURL)
+
+	// test install from remote domain
+	fileServerDir, err := ioutil.TempDir("", "sm_fileserver")
+	if err != nil {
+		t.Fatalf("Error create temporary dir: %s", err)
+	}
+
```

Comment:
Remove empty line. Or linter doesn't allow it?

Reply:
Linter detects as error